Worker logs WARN: Worker Evaluate Failed with error 'null' when a flow with realtime triggers is deleted

### Describe the issue

No WARN should be logged when a realtime trigger is stopped/killed.

### Environment

- Kestra Version: 0.17
- Operating System (OS/Docker/Kubernetes):
- Java Version (if you don't run kestra in Docker):

This is due to the fact that we kill the trigger but as we didn't wait the thread is still running and interrupted.

Sometimes it even display a stacktrace:
```
2024-06-04 17:17:27,311 WARN  worker_1     flow.realtime-pubsub-delete.trigger [namespace: company.myteam] [flow: realtime-pubsub-delete] [trigger: trigger] [date: 2024-06-04T17:17:18+02:00[Europe/Paris]] Worker Evaluate Failed with error 'java.lang.InterruptedException'
reactor.core.Exceptions$ReactiveException: java.lang.InterruptedException
	at reactor.core.Exceptions.propagate(Exceptions.java:410)
	at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:96)
	at reactor.core.publisher.Flux.blockLast(Flux.java:2817)
	at io.kestra.core.runners.WorkerTriggerRealtimeThread.doRun(WorkerTriggerRealtimeThread.java:41)
	at io.kestra.core.runners.AbstractWorkerThread.run(AbstractWorkerThread.java:56)
Caused by: java.lang.InterruptedException: null
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1100)
	at java.base/java.util.concurrent.CountDownLatch.await(CountDownLatch.java:230)
	at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:91)
	... 3 common frames omitted
```

A solution would be to add a little delay in the `AbstractWorkerTriggerThread` to wait for the thread to finish so it didn't interrupt itself.

--- a/core/src/main/java/io/kestra/core/runners/AbstractWorkerThread.java
+++ b/core/src/main/java/io/kestra/core/runners/AbstractWorkerThread.java
@@ -87,7 +87,12 @@ public abstract class AbstractWorkerThread extends Thread {
     protected void kill(boolean markAsKilled) {
         this.killed = markAsKilled;
         this.taskState = KILLED;
-        this.interrupt();
+
+        // When we arrive here, the thread run() method may be ended but the thread "in the stopping process".
+        // So we don't interrupt if the shutdownLatch is 0 as this means the run() method is done or if the thread is no more alive.
+        if (shutdownLatch.getCount() > 0 && this.isAlive()) {
+            this.interrupt();
+        }
     }
 
     protected void exceptionHandler(Thread t, Throwable e) {

--- a/core/src/main/java/io/kestra/core/runners/AbstractWorkerTriggerThread.java
+++ b/core/src/main/java/io/kestra/core/runners/AbstractWorkerTriggerThread.java
@@ -3,7 +3,12 @@ package io.kestra.core.runners;
 import io.kestra.core.models.triggers.WorkerTriggerInterface;
 import lombok.Getter;
 
+import java.time.Duration;
+
 public abstract class AbstractWorkerTriggerThread extends AbstractWorkerThread {
+    // This duration is by design low as we don't want to hang a thread for too long when we kill a trigger
+    private static final Duration AWAIT_ON_KILL = Duration.ofMillis(50);
+
     @Getter
     WorkerTrigger workerTrigger;
 
@@ -15,7 +20,7 @@ public abstract class AbstractWorkerTriggerThread extends AbstractWorkerThread {
     @Override
     public void signalStop() {
         try {
-            ((WorkerTriggerInterface)workerTrigger.getTrigger()).stop();
+            ((WorkerTriggerInterface) workerTrigger.getTrigger()).stop();
         } catch (Exception e) {
             logger.warn("Error while stopping trigger: '{}'", getType(), e);
         }
@@ -24,7 +29,13 @@ public abstract class AbstractWorkerTriggerThread extends AbstractWorkerThread {
     @Override
     protected void kill(final boolean markAsKilled) {
         try {
-            ((WorkerTriggerInterface)workerTrigger.getTrigger()).kill();
+            ((WorkerTriggerInterface) workerTrigger.getTrigger()).kill();
+            if (markAsKilled) {
+                // Let some time for the target thread to end, so we have a chance to not have to interrupt it.
+                // Killing a trigger is part of normal operations (updating a flow, disabling it, restarting Kestra),
+                // so we want to have a chance to end them properly and release their resources (transactions for ex).
+                awaitStop(AWAIT_ON_KILL);
+            }
         } catch (Exception e) {
             logger.warn("Error while killing trigger: '{}'", getType(), e);
         } finally {

--- a/core/src/main/java/io/kestra/core/runners/Worker.java
+++ b/core/src/main/java/io/kestra/core/runners/Worker.java
@@ -167,7 +167,6 @@ public class Worker implements Service, Runnable, AutoCloseable {
 
             synchronized (this) {
                 if (executionKilled.getLeft() instanceof ExecutionKilledExecution executionKilledExecution) {
-                    // @FIXME: the hashset will never expire killed execution
                     killedExecution.add(executionKilledExecution.getExecutionId());
 
                     workerThreadReferences
@@ -556,15 +555,25 @@ public class Worker implements Service, Runnable, AutoCloseable {
     private void logError(WorkerTrigger workerTrigger, Throwable e) {
         Logger logger = workerTrigger.getConditionContext().getRunContext().logger();
 
-        logService.logTrigger(
-            workerTrigger.getTriggerContext(),
-            logger,
-            Level.WARN,
-            "[date: {}] Worker Evaluate Failed with error '{}'",
-            workerTrigger.getTriggerContext().getDate(),
-            e != null ? e.getMessage() : "null",
-            e
-        );
+        if (e instanceof InterruptedException || (e != null && e.getCause() instanceof InterruptedException)) {
+            logService.logTrigger(
+                workerTrigger.getTriggerContext(),
+                logger,
+                Level.WARN,
+                "[date: {}] Trigger evaluation interrupted in the worker",
+                workerTrigger.getTriggerContext().getDate()
+            );
+        } else {
+            logService.logTrigger(
+                workerTrigger.getTriggerContext(),
+                logger,
+                Level.WARN,
+                "[date: {}] Trigger evaluation failed in the worker with error: {}",
+                workerTrigger.getTriggerContext().getDate(),
+                e != null ? e.getMessage() : "unknown",
+                e
+            );
+        }
 
         if (logger.isTraceEnabled() && e != null) {
             logger.trace(Throwables.getStackTraceAsString(e));
@@ -645,7 +654,7 @@ public class Worker implements Service, Runnable, AutoCloseable {
             workerThread.join();
             state = workerThread.getTaskState();
         } catch (InterruptedException e) {
-            logger.error("Failed to join WorkerThread {}", e.getMessage(), e);
+            logger.error("Failed to join the Worker thread: {}", e.getMessage(), e);
             if (workerThread instanceof WorkerTaskThread workerTaskThread) {
                 state = workerTaskThread.getWorkerTask().getTask().isAllowFailure() ? WARNING : FAILED;
             } else {
@@ -839,7 +848,7 @@ public class Worker implements Service, Runnable, AutoCloseable {
                 } else if (workerThread instanceof AbstractWorkerTriggerThread workerTriggerThread) {
                     return workerTriggerThread.workerTrigger;
                 } else {
-                    throw new Exception("Invalid thread type: '" + workerThread.getClass().getName() + "'");
+                    throw new IllegalArgumentException("Invalid thread type: '" + workerThread.getClass().getName() + "'");
                 }
             }))
             .toList();

--- a/jdbc/src/main/java/io/kestra/jdbc/runner/JdbcExecutor.java
+++ b/jdbc/src/main/java/io/kestra/jdbc/runner/JdbcExecutor.java
@@ -226,8 +226,8 @@ public class JdbcExecutor implements ExecutorInterface, Service {
 
                 try {
                     scheduledDelayFuture.get();
-                } catch (ExecutionException | InterruptedException e) {
-                    if (e.getCause().getClass() != CannotCreateTransactionException.class) {
+                } catch (ExecutionException | InterruptedException | CancellationException e) {
+                    if (e.getCause() != null && e.getCause().getClass() != CannotCreateTransactionException.class) {
                         log.error("Executor fatal exception in the scheduledDelay thread", e);
                         close();
                         KestraContext.getContext().shutdown();

